### PR TITLE
Fix: Add id_token grant flow

### PR DESCRIPTION
**Pull request by kangmingtay**
## What kind of change does this PR introduce?
Allows a user to send an id_token & a nonce via `POST /token?grant_type=id_token` to signup / signin a user

## Context
Some native mobile frameworks (such as Swift UI Authentication Services, Expo AuthSession) support their own existing OAuth2.0 authorization flow. The most common (and recommended) OAuth flow for native apps is the [Authorization Code flow with proof of key for code exchange](https://auth0.com/docs/flows/authorization-code-flow-with-proof-key-for-code-exchange-pkce). 
 
## What is the current behavior?
Currently, mobile developers that make use of their respective native frameworks for auth have difficulties integrating with gotrue. 

## What is the new behavior?
Mobile developers can use their own native frameworks for auth and send the id_token to gotrue via `POST /token?grant_type=id_token` endpoint to register a user in the auth schema. This partially addresses #140 and #198. 

## Example
```
POST /token?grant_type=id_token
{
  "id_token": "id_token_with_nonce",
  "nonce": "unhashed_nonce_in_id_token",
  "provider": "apple | google | azure | facebook"
}
```

## Limitations
Since the `id_token` is specific to the OpenID Connect (OIDC) standard, this new endpoint will only be able to support OIDC compliant providers for the meantime.   

## How the overall flow would look like
image

### CODE REVIEW

File: api/token.go [hunk 1]
```diff
@@ -44,9 +50,45 @@ type RefreshTokenGrantParams struct {
...
 const useSessionCookie = "session"
 
+func (p *IdTokenGrantParams) getVerifier(ctx context.Context) (*oidc.IDTokenVerifier, error) {
+	config := getConfig(ctx)
+	var provider *oidc.Provider
+	var err error
+	var clientId string
+	switch p.Provider {
+	case "apple":
+		clientId = config.External.Apple.ClientID
```
Comment (awalias) [inline api/token.go:70]: should we check if each provider is explicitly currently enabled in the config?

Comment (awalias) [reply api/token.go:70]: just in case someone wants to toggle it off

**Review by awalias — APPROVED** [review]
just a couple of small things, but overall looks ace

File: api/token.go (hunk 1 above)
Comment (kangmingtay) [reply api/token.go:70]: good idea! will add this

**Review by awalias — APPROVED** [review]
perfect! will leave you to merge it when you ready

**Comment by He1nr1chK** [comment]
Hi @kangmingtay, any update on when this will get merged? Super keen for this feature.

**Comment by libovness** [comment]
Same here!

**Comment by kangmingtay** [comment]
Hey @He1nr1chK , @libovness, we are working on a refactor [here](https://github.com/supabase/gotrue/pull/208) that will fix a couple of issues with the "Sign-up with an external OAuth provider" flow so this PR is unlikely to be merged yet. Some of the issues we aim to fix before introducing more sign-up flows (like this one) would be the following:
1. Sign-ups via an external provider relies on email as the primary identifier / OAuth fails when no email is linked to provider account #214 
2. Updating the email associated with the provider results in loss of access to user on gotrue
3. Only the first provider's metadata is stored, and subsequent sign-ins with other providers are not 

In term of timeline, I'm hoping to get the refactoring work done by this week and make the necessary changes to this PR next week. 

Thanks for understanding!

**Comment by libovness** [comment]
hi @kangmingtay checking in to see if the PR will happen this week?

**Comment by 10000multiplier** [comment]
Looks very exciting, crossing fingers for a quick merge 🤞

**Comment by thecoolwinter** [comment]
Any updates/timeline? I'm excited to integrate these changes with the iOS libraries so we can support Sign in With Apple

**Comment by kangmingtay** [comment]
Hey @libovness, @thecoolwinter, @10000multiplier, @He1nr1chK, @k0shk0sh, apologies for the delay, i'm still testing out this feature on my end and have some questions regarding the auth flow for ios / android:
* For android, is the android package name globally unique? (similar to the bundle identifier for ios apps)
* For external oauth authentication (sign-in with google) for android, I'm assuming the auth flow depends on the native mobile sdk / framework being used, but will one always get back an `id_token` at the end?
* If so, will the audience field in the `id_token` be set to the android package name during the auth flow?  

Once again, apologies for the delay in getting this out. Was pretty busy with the refactoring work in the last month and just got back from a trip this week 😅 Thanks so much for the patience so far!

**Comment by thecoolwinter** [comment]
For iOS the App ID and Service ID are separate things, but are linkable. 

To set up SWIA I would go and make a new app and set its Bundle ID, then create a Service ID. When creating that Service ID I'll actually link the Service ID to the bundle ID that I just created. So the Service ID can be used for the same mobile and web app.

Actually, the Service ID can be linked to any number of Bundle IDs or web IDs by grouping their identifiers so that won't be a problem at all.

Thank you for all your hard work 🙏 this is awesome!

**Comment by kangmingtay** [comment]
Hi everyone, it's been a long wait and i'm happy to announce that we'll be merging this feature into gotrue, and releasing it to all supabase projects in the upcoming week. Do note that this endpoint is still in alpha and thus susceptible to changes. Thanks to everyone for the patience and valuable insights in terms of helping me understand the issues the mobile community has been facing so far!

**Comment by himeshp** [comment]
hi, @kangmingtay does this require updating the supabase dart library to access the end point?

**Comment by bdlukaa** [comment]
@dshukertjr

**Comment by dshukertjr** [comment]
@bdlukaa Thanks for the heads up! This is exciting!

@himeshp Yup, this will require us to update the `supabase_dart` library! Looks like this will be the next big update in `supabase_dart`!

**Pull request merged**